### DISCUSSION TITLE
Million Nodes creation

### DISCUSSION THREAD

**Discussion by jigar88** [FreeOpcUa/python-opcua#1302] [Ideas]
Hello,

I am developing OPC server using this library. However my use case requires a lot of variable nodes to create. Is it possible to create Million nodes quickly. Currently even with threading server response is very slow and it takes forever to create that many nodes. it creates roughly 4000 nodes per minute which is quite slow for my use case. Any suggestion ?

**Comment by swamper123** [comment]
I guess the fastest way to get nodes is loading a pickle file, you dumped before. 

https://github.com/FreeOpcUa/python-opcua/blob/858274721766907148b527916e97446dbbf6af6d/opcua/server/internal_server.py#L159
Remember: Never touch pickles from strangers, unless you realy trust the file.

Less faster would be creating on you own in Python (which you may have done) and the least is via xml. Perhaps a bit execution boost may happen if you compile your Pythoninterpreter with --enable-optimisations...but if you realy need it even more faster, Python may not be your programming language of choice.

**Comment by starturtle** [comment]
Please calculate your required resources.
A million nodes, if created at 1ms per node (roughly 20 times faster than what you perceived), still take over 15 min to create.
Maybe some very custom trickery (that emulates rather than creates nodes) is closer to your use case than the default implementation. Also, like swamper mentioned, python may not be for you.
Good luck!

**Comment by zerox1212** [comment]
Python won't be great for this. Even OPCUA isn't good for this because it's a very heavy protocol.

You could see if PyPy is faster.

Otherwise you probably need to rethink your implementation... do you really need one million nodes up front? Why not create as needed or some such thing.

**Reply by jigar88** [reply]
@zerox1212 . I am creating these nodes as they come or require to create. So I have my application running on aws ec2 instance. The problem is when we test fail over or when I restarted the ec2 server it will take way long to recreate all those nodes. Is there a way to export all nodes and import after server come back online?

**Reply by zerox1212** [reply]
You can try dumping the address space, but honestly I have no idea how it will handle dumping so many nodes.
